test/functional/p2p_sendrecon.py

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/env python3
+# Copyright (c) 2022 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test SENDRECON message
+"""
+
+import time
+
+from test_framework.messages import (
+    msg_sendrecon,
+    msg_verack,
+    msg_version,
+)
+from test_framework.p2p import (
+    P2PInterface,
+    P2P_SERVICES,
+    P2P_SUBVERSION,
+)
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+class SendReconSender(P2PInterface):
+    def __init__(self, manual_verack=False):
+        super().__init__()
+        self.version_received = False
+        self.sendrecon_msg_received = None
+        self.manual_verack = manual_verack
+
+    def on_version(self, message):
+        if not self.manual_verack:
+            super().on_version(message)
+        self.version_received = True
+
+    def on_sendrecon(self, message):
+        print(message)
+        self.sendrecon_msg_received = message
+
+def create_sendrecon_msg():
+    sendrecon_msg = msg_sendrecon()
+    sendrecon_msg.initiator = True
+    sendrecon_msg.responder = False
+    sendrecon_msg.version = 1
+    sendrecon_msg.salt = 2
+    return sendrecon_msg
+
+class SendReconTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.extra_args = [['-txrecon']]
+
+    def run_test(self):
+        # Checks for the node *sending* SENDRECON
+        self.log.info('SENDRECON sent')
+        # Inbound
+        peer = self.nodes[0].add_p2p_connection(SendReconSender(), send_version=True, wait_for_verack=False)
+        peer.wait_until(lambda: peer.sendrecon_msg_received)
+        assert not peer.sendrecon_msg_received.initiator
+        assert peer.sendrecon_msg_received.responder
+        assert_equal(peer.sendrecon_msg_received.version, 1)
+        # Outbound
+        peer = self.nodes[0].add_outbound_p2p_connection(SendReconSender(), p2p_idx=1, connection_type="outbound-full-relay")
+        peer.wait_until(lambda: peer.sendrecon_msg_received)
+        assert peer.sendrecon_msg_received.initiator
+        assert not peer.sendrecon_msg_received.responder
+        assert_equal(peer.sendrecon_msg_received.version, 1)
+
+        self.log.info('SENDRECON should not be sent if block-relay-only')
+        peer = self.nodes[0].add_outbound_p2p_connection(SendReconSender(), p2p_idx=2, connection_type="block-relay-only")
+        time.sleep(1)
+        assert not peer.sendrecon_msg_received
```

Comment:
you want to `sync_with_ping()` here, and then assert not received. otherwise the message could be unreceived due to race condition

Reply:
This doesn't work prior to verack? Where it's possible, i add it. Here, i can increase a timer... or what would you suggest?

Reply:
ah my bad! well, could you wait until verack handshake is done, and then assert that sendrecon was never recieved?